artemis-protocols/artemis-amqp-protocol/src/test/java/org/apache/activemq/artemis/protocol/amqp/sasl/SCRAMTest.java

```diff
@@ -0,0 +1,192 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.activemq.artemis.protocol.amqp.sasl;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertTrue;
+
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.ArrayList;
+import java.util.List;
+
+import javax.crypto.Mac;
+import javax.security.auth.login.LoginException;
+
+import org.apache.activemq.artemis.protocol.amqp.sasl.scram.SCRAMClientSASL;
+import org.apache.activemq.artemis.protocol.amqp.sasl.scram.SCRAMServerSASL;
+import org.apache.activemq.artemis.protocol.amqp.sasl.scram.ScramServerFunctionalityImpl;
+import org.apache.activemq.artemis.spi.core.security.scram.SCRAM;
+import org.apache.activemq.artemis.spi.core.security.scram.ScramException;
+import org.apache.activemq.artemis.spi.core.security.scram.ScramUtils;
+import org.apache.activemq.artemis.spi.core.security.scram.UserData;
+import org.apache.qpid.proton.codec.DecodeException;
+import org.hamcrest.core.IsInstanceOf;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameters;
+
+/**
+ * test cases for the SASL-SCRAM
+ */
+@RunWith(Parameterized.class)
+public class SCRAMTest {
+
+   /**
+    *
+    */
+   private final SCRAM mechanism;
+   private static final byte[] SALT = new byte[32];
+   private static final String SNONCE = "server";
+   private static final String CNONCE = "client";
+   private static final String USERNAME = "test";
+   private static final String PASSWORD = "123";
+
+   @Parameters(name = "{0}")
+   public static List<Object[]> data() {
+      List<Object[]> list = new ArrayList<>();
+      for (SCRAM scram : SCRAM.values()) {
+         list.add(new Object[] {scram});
+      }
+      return list;
+   }
+
+   public SCRAMTest(SCRAM mechanism) {
+      this.mechanism = mechanism;
+   }
+
+   @Test
+   public void testSuccess() throws NoSuchAlgorithmException {
+      TestSCRAMServerSASL serverSASL = new TestSCRAMServerSASL(mechanism, USERNAME, PASSWORD);
+      TestSCRAMClientSASL clientSASL = new TestSCRAMClientSASL(mechanism, USERNAME, PASSWORD);
```

Comment:
So this type of test indeed will rarely give a different result than the integration test, since its using mostly the same code doing the same thing, I can see why you would wonder about having both with this test in particular. 

Wiring them in a circle like this doesnt really verify the behaviour is truly as expected, so much as that neither side blew up. I'd expect each side to be tested in isolation with some actual verification of its generated output.

E.g for the client side, https://github.com/apache/qpid-jms/tree/main/qpid-jms-client/src/test/java/org/apache/qpid/jms/sasl contains example of the kind of client handling tests I would expect. At least for the SHA1 and 256 cases for example, it uses the example straight from the RFC to check the actual behaviour based on known input.

Reply:
> Wiring them in a circle like this doesnt really verify the behaviour is truly as expected
it at least verifies that the client+server side could work together. combined with the QPID test it verifies that other implementations behave similar.

> it uses the example straight from the RFC to check the actual behaviour based on known input.
I could add those as well but this also jut test one fixed set of data. So even if my test-string are not "visible" in code they always use the same string.